Mathlib/GameTheory/Auction/Basic.lean

```diff
@@ -0,0 +1,193 @@
+  rw [bid_winner_eq_maxBid b]
+  exact Finset.le_sup' b (Finset.mem_univ j)
+
+/-- If `i`'s bid is higher than all other bids, then `i` is the winner. -/
+lemma eq_winner_of_bid_gt (i : I) (H : ∀ j , j ≠ i → b j < b i) : i = winner b := by
+  contrapose! H
+  exact ⟨winner b, H.symm, bid_le_bid_winner b i⟩
+
+variable [DecidableEq I]
+
+/-- A strategy is dominant if bidding `bi` ensures that
+`i`'s utility is maximized relative to any other bids `b'` where `b i = bi`. -/
+def Dominant (utility : (I → ℝ) → I → ℝ) (i : I) (bi : ℝ) : Prop :=
+  ∀ b, utility b i ≤ utility (Function.update b i bi) i
+
+/-! ### Proofs and Lemmas -/
+
+namespace Firstprice
+
+/-- Computes the utility for a first price auction where the winner pays their bid. -/
+@[simp]
+noncomputable def utility [DecidableEq I] (v b : I → ℝ) (i : I) : ℝ :=
+    if i = winner b then v i - b i else 0
+
+/-- If `i` is the winner in a first price auction, utility is their valuation minus their bid. -/
+lemma utility_winner (i : I) (H : i = winner b) : utility v b i = v i - b i := by
+  rw [H]
+  simp only [utility, if_true]
+
+/-- If `i` is not the winner in a first price auction, their utility is 0. -/
+lemma utility_loser (i : I) (H : i ≠ winner b) : utility v b i = 0 := by
+  rw [utility]
+  simp only [H, if_false]
+
+/-- Shows that there is no dominant strategy in a first price auction for any `i` and bid `bi`. -/
+theorem firstprice_auction_has_no_dominant_strategy (i : I) (bi : ℝ) :
+    ¬ Dominant (utility v) i bi := by
+  rw [Dominant, not_forall]
+  use Function.update (fun _ ↦ (bi - 2)) i (bi - 1)
+  rw [utility_winner _ _ _ (eq_winner_of_bid_gt _ i _),
+      utility_winner _ _ _ (eq_winner_of_bid_gt _ i _)]
+  <;> intros <;> simp [*]
+
+end Firstprice
+
+/-- `maxBidExcluding i` is the maximal bid of all participants but `i`. -/
+noncomputable def maxBidExcluding (i : I) : ℝ :=
+  (Finset.univ.erase i).sup' (Finset.univ_nontrivial.erase_nonempty) b
+
+/--The second highest bid: the highest bid excluding the winner’s bid.-/
+noncomputable def secondPrice : ℝ := maxBidExcluding b (winner b)
+
+namespace Secondprice
+
+/-- The bid of the winner is always greater than or equal to the second highest bid. -/
+lemma maxBidExcluding_le_maxBid {i : I} (b : I → ℝ) : maxBidExcluding b i ≤ maxBid b := by
+  apply Finset.sup'_mono
+  exact Finset.subset_univ _
+
+/-- If `i` is not the winner, then the highest bid excluding `i` is equal to the highest bid. -/
+lemma maxBidExcluding_eq_maxBid_if_loser {i : I} (H : i ≠ winner b) :
+    maxBidExcluding b i = maxBid b := by
+  apply le_antisymm
+  · exact maxBidExcluding_le_maxBid b
+  · rw [← bid_winner_eq_maxBid b]
+    apply Finset.le_sup'
+    exact Finset.mem_erase_of_ne_of_mem H.symm (Finset.mem_univ (winner b))
+
+/-- Defines the utility of participant `i`,
+which is their valuation minus the second highest bid if `i` is the winner, otherwise, it's 0. -/
+noncomputable def utility [DecidableEq I] (v b : I → ℝ) (i : I) : ℝ :=
```

Comment:
Likewise, is this necessary?
```suggestion
noncomputable def utility (v b : I → ℝ) (i : I) : ℝ :=
```

Reply:
Can drop the `noncomputable` too I think